src/components/patterns/on-the-page.tsx

```diff
@@ -0,0 +1,136 @@
+"use client"
+
+import React, {useState, useEffect} from "react"
+import {NodeStanfordNews, NodeStanfordPage, NodeSulLibrary} from "@/lib/gql/__generated__/drupal.d"
+import {twMerge} from "tailwind-merge"
+import DrupalLink from "./elements/drupal-link"
+import SelectList from "./elements/select-list"
+import {SelectOptionDefinition} from "@mui/base/useSelect"
+
+interface Heading {
+  text: string
+  id: string
+}
+
+const OnThePageLink = ({node}: {node: NodeStanfordPage | NodeStanfordNews | NodeSulLibrary}) => {
+  const [headings, setHeadings] = useState<Heading[]>([])
+  const [activeHeading, setActiveHeading] = useState<string>("")
+
+  const relLinkHeading = "sulRelLinksHeading" in node ? node.sulRelLinksHeading : undefined
+  const relLinks = "sulRelLinks" in node ? node.sulRelLinks : undefined
+
+  useEffect(() => {
+    const wysiwygContainers: NodeListOf<Element> = document.querySelectorAll(".wysiwyg")
+
+    const allHeadings: Heading[] = []
+
+    wysiwygContainers.forEach((container, containerIndex) => {
+      const h2Elements: HTMLHeadingElement[] = Array.from(container.querySelectorAll("h2"))
+
+      h2Elements.forEach((heading, headingIndex) => {
+        const id = `${heading.textContent?.trim().toLowerCase().replace(/\s+/g, "-")}-${containerIndex}-${headingIndex}`
+        heading.setAttribute("id", id)
+        allHeadings.push({text: heading.textContent || "", id})
+      })
+    })
+
+    setHeadings(allHeadings)
+
+    const observerCallback: IntersectionObserverCallback = entries => {
+      entries.forEach(entry => {
+        if (entry.isIntersecting && entry.intersectionRatio > 0) {
+          setActiveHeading(entry.target.id)
+        }
+      })
+    }
+
+    const observer = new IntersectionObserver(observerCallback, {
+      rootMargin: "-10px 0px 0px 0px",
+      threshold: 0,
+    })
+
+    wysiwygContainers.forEach(container => {
+      const h2Elements = container.querySelectorAll("h2")
+      h2Elements.forEach(heading => observer.observe(heading))
+    })
+
+    return () => {
+      observer.disconnect()
+    }
+  }, [])
+
+  const options: SelectOptionDefinition<string>[] = [
+    ...headings.map(heading => ({
+      value: heading.id,
+      label: heading.text,
+    })),
+    {
+      value: "related-links-header",
+      label: node.sulRelLinksHeading || "Related content",
+      disabled: true,
+    },
+    ...(node.sulRelLinks?.map(link => ({
+      value: String(link.url),
+      label: String(link.title),
+    })) || []),
+  ]
+
+  return (
+    <>
+      <div className="block w-full md:w-500 lg:hidden">
+        <SelectList
+          label="On the page"
+          options={options}
+          onChange={(event, value) => {
+            if (value) {
+              const selectedHeading = document.getElementById(value as string)
+              if (selectedHeading) {
+                selectedHeading.scrollIntoView({behavior: "smooth"})
```

Comment:
also add `block: "end", inline: "nearest"` options. i find that to be better behavior.